### DISCUSSION TITLE
Your main branch isn't protected

### DISCUSSION THREAD

**Discussion by OverLordGoldDragon** [community/community#23046] [New to GitHub] [answered]
<p>After running a bunch of commands trying to create and delete tags and branches, I see the following</p>
<p><a href="https://i.stack.imgur.com/WT8wk.png" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/16495490/181074529-3d4ef630-bed8-442e-84ed-1176235b0945.png" alt="enter image description here" width="690" height="103"></a></p>
<p>What is it, and how should I handle it? My other repositories don’t have it.</p>

**Comment by jsoref** [comment]
<p>This appears to be a <em>nudge</em> that GitHub is incrementally deploying.</p>
<p>The best practice is to add protection.</p>
<p><a href="https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches" class="onebox" target="_blank" rel="noopener">https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches</a></p>

**Comment by GLD5000** [comment]
<p>Ditto- it is just a reminder to set some rules on your main branch - nothing to worry about!</p>
<p>You can just set the rule to disallow direct pushing to main, for example, and it will go away.</p>
<p>Hope that helps,</p>
<p>G</p>

**Comment by OverLordGoldDragon** [comment]
<p>Thanks. So what is the trigger? I did just notice that some of my other repositories have it, but can’t tell what sets them apart.</p>

**Comment by jsoref** [comment]
<p>I doubt anyone from GitHub will answer. Large entities do A-B testing, canary releases, incremental releases, etc.</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://harness.io/blog/continuous-verification/blue-green-canary-deployment-strategies/">
  <header class="source">
      <img src="https://user-images.githubusercontent.com/2119212/181074532-0e31e2b9-1f41-4835-99e4-2ec6c735caa2.png" class="site-icon" width="80" height="80">

      <a href="https://harness.io/blog/continuous-verification/blue-green-canary-deployment-strategies/" target="_blank" rel="noopener" title="05:00PM - 15 January 2021">Harness.io – 15 Jan 21</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/215;"><img src="https://user-images.githubusercontent.com/2119212/181074533-73bd5ad2-36ee-4bcf-bb0e-e32e69a0b51f.jpeg" class="thumbnail" width="690" height="215" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/9/4/94d0d1e261fc958ede53512935407cb495c20460_2_690x215.jpeg, https://aws1.discourse-cdn.com/github/optimized/3X/9/4/94d0d1e261fc958ede53512935407cb495c20460_2_1035x322.jpeg 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/9/4/94d0d1e261fc958ede53512935407cb495c20460_2_1380x430.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/9/4/94d0d1e261fc958ede53512935407cb495c20460_2_10x10.png"></div>

<h3><a href="https://harness.io/blog/continuous-verification/blue-green-canary-deployment-strategies/" target="_blank" rel="noopener">Intro to Deployment Strategies: Blue-Green, Canary, and More | Harness</a></h3>

  <p>In our Blue-Green vs Canary deployment strategies blog, we go over basic, rolling, blue-green, &amp; canary to help you make the right choice.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>It’s possible they randomly selected one repository per user for some fraction of their user base. There’s no reason to do it for all your repositories, that’d presumably annoy you more, which isn’t what they need when they’re gathering data.</p>
<p>Anyway, it is a best practice, so you could just take the hint and apply it everywhere (you don’t have to, I haven’t, although I probably will for the base repositories that I care about).</p>

**Comment by OverLordGoldDragon** [comment]
<p>Well if everyone’s being pinged to ‘fix’ something, at the least a brief explanation is due so we don’t think we broke something. I hope they add some info or reply here.</p>

**Comment by ryanwags** [comment]
<p>I thought this feature was only available for paid tiers or public repos. That’s what their pricing page still says. I couldn’t find any announcement about Branch Protections being made available to everyone, did I miss something?</p>

**Comment by jsoref** [comment]
<p>If you want a direct response, open a ticket via:</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://support.github.com/">
  <header class="source">
      <img src="https://user-images.githubusercontent.com/2119212/181074534-2e828f49-0765-4858-bc51-24496b818e18.png" class="site-icon" width="32" height="32">

      <a href="https://support.github.com/" target="_blank" rel="noopener">support.github.com</a>
  </header>

  <article class="onebox-body">
    

<h3><a href="https://support.github.com/" target="_blank" rel="noopener">GitHub Support</a></h3>

  <p>GitHub Support is here to help. Learn about GitHub products, browse our helpful resources, and contact support with your questions.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Don’t expect an instant response (it could take 2 weeks, especially if they happen to take a company-wide vacation for a week, but that happened recently, so I don’t expect it to happen again anytime soon.)</p>

**Comment by jsoref** [comment]
<p>I don’t think anyone has said they’re an unpaid-subscription w/ a private repo.</p>

**Comment by ryanwags** [comment]
<p>Sorry, I know OP didn’t say they were unpaid/private. I came here looking for info about the same topic and this was the only post I found on the matter.</p>

**Comment by realmazharhussain** [comment]
<p>This appeared on one of my repos when I pushed some commits to main branch and then did a <code>git reset</code> on local repo and then force pushed again.</p>
<p>It seems like this message appears when we force push something to the main branch.</p>

**Comment by ryanckulp** [comment]
<p>perhaps just tangentially related but we have a free Github plugin that adds more granular controls to protected branches, merges, and code freezes.</p>
<p>this also lets you add a “note” whenever a branch gets frozen, so other team members have context.</p>
<p>100% free for open source:</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/marketplace/merge-freeze">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/marketplace/merge-freeze" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/3083888/181074538-d35d1536-0166-455d-beaf-96ee9ba59773.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/c/9/c9398ee325b749217325115a094ce359978ecea5_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/c/9/c9398ee325b749217325115a094ce359978ecea5_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/c/9/c9398ee325b749217325115a094ce359978ecea5.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/c/9/c9398ee325b749217325115a094ce359978ecea5_2_10x10.png"></div>

<h3><a href="https://github.com/marketplace/merge-freeze" target="_blank" rel="noopener nofollow ugc">Merge Freeze - GitHub Marketplace</a></h3>

  <p>The code freeze tool to block merging and deployments</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Ryan</p>

**Comment by OverLordGoldDragon** [comment]
GitHub Support:

 > Please do not be alarmed, you have not broken anything on your repository. We introduced a prompt for repository owners and admins to protect their default branches ... the prompt [is] displayed to any admins with more than one branch in a repository.

I've confirmed this, creating branch creates banner, deleting branch deletes banner.

**Comment by fransf-wtax** [comment]
<p>I think you’ve hit the nail on the head: this is obviously a way GitHub has come up with to scare people who need to keep their repositories private into upgrading to a paid tier.</p>

**Comment by Alleged627** [comment]
<p>Same thing is happening to me</p>

**Comment by brandonrosage** [comment, marked as answer]
👋🏼 Thanks for the feedback, y'all. I'm Brandon Rosage, a product manager at GitHub, and I'm leading an initiative that introduced these messages about branch protection rules.

The intent of introducing the message was to make the feature -- the ability to create rules that protect your branches -- more discoverable and accessible to all users. This intent was derived from feedback that users who use "protected branches" gave us regarding how highly they value it, despite having to work hard to learn about and/or stumble upon it in the first place.

So we developed an initiative that introduces an experiment in which admins for any repo that has more than one branch and no existing branch protection rules is told that their default branch isn't protected, with a button to add a rule. 

In doing so, we also introduced access to creating branch protection rules on any repo. Before, it was inaccessible to private repos owned by accounts using GitHub Free. With the introduction of the experiment, however, those repos are limited only in the _enforcement_ of the rules they create.

We make an earnest attempt to communicate all of this as users encounter the messaging. But the feedback y'all have provided here is gold. So we'll be iterating this month to ensure the most effective treatment is used, and that we don't stoke any fear that something is broken.

Rest assured, our intent is not to scare anyone (neither that anything is broken or that we prefer repos use a certain visibility setting). It's simply to communicate the value of a feature that users have told us they value, but had a hard time discovering.

I'm happy to address any other concerns or questions here 👍🏼

**Reply by jsoref** [reply]
Hi @brandonrosage: my general feedback is that whenever you introduce a nudge, you should include a blog link (or documentation link) that talks about the nudge.

I personally am quite tolerant of such things, but if your audience isn't going to expect it, then having a thing that explains what you're doing, why you're doing it, and the scope of people likely to encounter it will help everyone.

I understand that such tasks require more effort and thus increase friction before release, but they really make for a much smoother process.

**Reply by brandonrosage** [reply]
💯 👍🏼 We expect we'll have experiment results to work with in a couple weeks, after which we'll take everything we learned and bake it into the experience. Including links to more detail (e.g. docs) will absolutely be part of that, thanks to you.

**Reply by OverLordGoldDragon** [reply]
### TL;DR version

> We introduced a prompt for repository owners and admins to protect their default branches ... the prompt [is] displayed to any admins with more than one branch in a repository.

Thanks for clarifying.

**Reply by airjp73** [reply]
Thanks for iterating on this. I wanted to add that, I got this message on a repository where I'm the only person who has access to make changes directly, but I do get PRs now and then. Seeing the message was pretty confusing -- making me think maybe a contributor could just start pushing changes if they wanted unless I set up these rules. I definitely think branch protection rules can be useful even in my case, but I think the messaging should clearly explain that.

**Reply by Alleged627** [reply]
I give up none of this makes sense

**Reply by brandonrosage** [reply]
@Alleged627 Is there something I can help clarify?

**Reply by DavidCSchneider** [reply]
Hello Brandon,
This experiment really messed up my life.  I haven't a clue how to undo the branch protection so that I can restore the damage caused by changing a branch name, and then invoking the rule.  
@DavidCSchneider

**Reply by brandonrosage** [reply]
@DavidCSchneider Oh, no! I'm sincerely sorry this experiment created problems for you. I'd be happy to talk through your issue and troubleshoot it. Maybe invite me to the affected repository, or connect with me via email: `brandonrosage@github.com`.

**Reply by teacher-isaac** [reply]
@brandonrosage Thanks for the explanation. I agree with @jsoref. I think integrated/inline documentation with new features can help people navigate the complexity (and hopefully keep the documentation and features in sync).

**Reply by Alleged627** [reply]
Can you also help me figure out how I am supposed to protect it

**Reply by jsoref** [reply]
@Alleged627: For each repository, follow these instructions:
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/managing-a-branch-protection-rule#creating-a-branch-protection-rule

Personally, I tend to use:
- [x] Require status checks to pass before merging
- [x] Restrict who can push to matching branches

For teams, we tend to use:
- [x] Require a pull request before merging
   - [x] Require approvals

**Reply by Alleged627** [reply]
Thank you so much.  It’s not going to mess anything up by doing this right

**Reply by jsoref** [reply]
It shouldn't. Obviously you'll want to read the text and think through what each toggle means. If you run into problems, you can ask for support here (or via support).

If you're the only contributor to a repository, then requiring approvals will probably not work well because you can't approve your own PR.

**Reply by Alleged627** [reply]
Crap what do I do there was someone helping in the beginning I just didn’t know if it was good guys or bad guys at the time because I got hacked and so I started thinking everyone was trying to get me so I apologize if I messed stuff up

**Reply by ardunster** [reply]
I haven't used github for personal projects in a few months, and now all of my *personal* repos have this alert. it's super annoying. I don't need branch protection from myself, and I know it exists and how to use it when I want to. Can I disable this alert?

**Reply by mrleblanc101** [reply]
Would it be possible to remove the warning if there is no other contributors ?
I don't need to protect branches on project I'm alone.

**Reply by frauikram** [reply]
@brandonrosage Thank you for the explanation, however, instead of 
### **`your <main> branch isn't protected`**,
a text like:
### `No additional protection rules are configured for this branch` 
would have been helpful, in my honest opinion.

**Reply by pipedreambomb** [reply]
I think it's a valuable warning to give, but I think the follow-up is a bit lacking. When you click "View Documentation", you get the general purpose article, [About rulesets](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/about-rulesets), which is fine but doesn't tell the lay user anything about how to protect their main/master branch specifically. 

I think a specific suggested ruleset, perhaps a basic ruleset template even, would go so much further to help the unintiated with getting started with branch protection. There are so many options on the new ruleset page that it's pretty baffling. I've had a go but I don't know if what I did was good. I don't even know how to know if it is.

**Reply by blixten85** [reply]
> 👋🏼 Thanks for the feedback, y'all. I'm Brandon Rosage, a product manager at GitHub, and I'm leading an initiative that introduced these messages about branch protection rules.
> 
> The intent of introducing the message was to make the feature -- the ability to create rules that protect your branches -- more discoverable and accessible to all users. This intent was derived from feedback that users who use "protected branches" gave us regarding how highly they value it, despite having to work hard to learn about and/or stumble upon it in the first place.
> 
> So we developed an initiative that introduces an experiment in which admins for any repo that has more than one branch and no existing branch protection rules is told that their default branch isn't protected, with a button to add a rule.
> 
> In doing so, we also introduced access to creating branch protection rules on any repo. Before, it was inaccessible to private repos owned by accounts using GitHub Free. With the introduction of the experiment, however, those repos are limited only in the _enforcement_ of the rules they create.
> 
> We make an earnest attempt to communicate all of this as users encounter the messaging. But the feedback y'all have provided here is gold. So we'll be iterating this month to ensure the most effective treatment is used, and that we don't stoke any fear that something is broken.
> 
> Rest assured, our intent is not to scare anyone (neither that anything is broken or that we prefer repos use a certain visibility setting). It's simply to communicate the value of a feature that users have told us they value, but had a hard time discovering.
> 
> I'm happy to address any other concerns or questions here 👍🏼

More automation and templates would be great. For example, predefined repository security levels with preset rules: Minimal, Medium, and High. Users who want more control could still customize everything themselves.

The same applies to workflows, Dependabot, Renovate, CodeQL, and similar tools.

Why should I, as the end user, have to spend time setting these things up or even learning how they work? While they may be considered best practices, from my perspective they’re mostly a waste of time and energy.

What most people really want is a golden-standard repository setup that’s automatically tailored to the repository’s needs. A setup wizard or template creator would be ideal, allowing users to choose different security or automation levels while automatically detecting which tools, workflows, and packages are appropriate for the project.

I don’t know how to code, and I have no interest in learning. What I enjoy is building apps and websites, so AI has been a true blessing.

The problem is that I still have to spend time—and AI tokens—just to get my repository into what should already be considered a golden-standard state.

Things like renewing GitHub tokens or remembering where and when a token needs to be updated shouldn’t be my responsibility. The platform should handle that automatically behind the scenes whenever possible.

There also shouldn’t be unnecessary tokens left lying around. Credentials should be managed automatically, securely, and with the principle of least privilege by default.

Best regards
Anders
